### PR TITLE
150 name of place vs longlat

### LINKED ISSUE #150: Allow user to enter name of place instead of lat/lon in LatLonSetInput
### Description
Entering lat/lon manually is slightly impractical. The user should be able to search for a location and have the lat/lon autofilled.

### Subtasks
- [x] Select a mature and solid framework (Google? Mapbox? Something else entirely?)
- [x] Implement lat/lon search. Should be able to keep the changes isolated within the existing component so as not to disturb the interfaces &c

### PR DESCRIPTION
**Pull request by Kpaubert**
Add input field using GeoNorges' API to find longlat from an address.

Closes #150

### CODE REVIEW

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            return {
+                label: `${address.adressetekst} (${address.kommunenavn})`,
+                lat: address.representasjonspunkt.lat,
+                lon: address.representasjonspunkt.lon,
+            };
+        });
+        setLocations(locationList);
+    };
+
+    const inputChanged = (address: { label: string; lat: number; lon: number }): void => {
```
Comment (Kpaubert) [inline src/components/atoms/LatLonFromLocationSearch.tsx]: Rename to selectionChanged

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            return;
+        }
+        const locationList = locationData.adresser.map((address: Address) => {
+            return {
+                label: `${address.adressetekst} (${address.kommunenavn})`,
+                lat: address.representasjonspunkt.lat,
+                lon: address.representasjonspunkt.lon,
+            };
+        });
+        setLocations(locationList);
```
Comment (Kpaubert) [inline src/components/atoms/LatLonFromLocationSearch.tsx]: Could remove the first 4 lines, and rather have it be:
```ts
const locationList = locationDatal.adresser.length > 0 : locationData.adresser.map((address: Address) => {
            return {
                label: `${address.adressetekst} (${address.kommunenavn})`,
                lat: address.representasjonspunkt.lat,
                lon: address.representasjonspunkt.lon,
            };
        }) : [];
```
or something

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    setState: Dispatch<SetStateAction<MetAPIVariables>>;
+};
+
+const LatLonFromLocationSearch: React.FC<LatLonInputSetProps> = (props) => {
+    const [state, setState] = useState<{ lat: string; lon: string }>({
+        lat: String(props.state.lat),
+        lon: String(props.state.lon),
+    });
+
+    const [locations, setLocations] = useState<string[]>([]);
```
Comment (Kpaubert) [inline src/components/atoms/LatLonFromLocationSearch.tsx]: This type is not correct after changing it to be an object

**Review by Kpaubert — COMMENTED** [review]
Nice work, KP!

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            (position) => {
+                setState({ lat: String(position.coords.latitude), lon: String(position.coords.longitude) });
+            },
+            () => {
+                toaster.warning('Du må tillate lokasjonstjenester for at denne tjenesten skal fungere.');
+            },
+        );
+    };
+    // TODO: This is a SyntheticBaseEvent, don't know how to import that type. Any ideas, August?
+    const onInputChange = async (e: any): Promise<void> => {
```
Comment (fredrikbw) [inline src/components/atoms/LatLonFromLocationSearch.tsx]:
```suggestion
    const onInputChange = async (e: React.ChangeEvent<HTMLInputElement>): Promise<void> => {
```

Comment (Kpaubert) [reply src/components/atoms/LatLonFromLocationSearch.tsx]: Wow, nice catch! With this one, we're going to the moon! 🚀  *guitar solo* 🎸

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        lat: String(props.state.lat),
+        lon: String(props.state.lon),
+    });
+
+    const [locations, setLocations] = useState<
+        {
+            label: string;
+            lat: number;
+            lon: number;
+        }[]
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]: Move this out to its own type.

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            () => {
+                toaster.warning('Du må tillate lokasjonstjenester for at denne tjenesten skal fungere.');
+            },
+        );
+    };
+
+    const onInputChange = async (e: React.ChangeEvent<HTMLInputElement>): Promise<void> => {
+        const locationData = await fetch(
+            `https://ws.geonorge.no/adresser/v1/sok?sok=${e.target.value}&treffPerSide=10`,
+        ).then((res) => res.json());
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]:
```suggestion
        const rawLocationData = await fetch(
            `https://ws.geonorge.no/adresser/v1/sok?sok=${e.target.value}&treffPerSide=10`,
        );
        const locationData: Array<Address> = await rawLocationData.json();
```

Use either Promises or async/await.

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        const locationList =
+            locationData.adresser.length > 0
+                ? locationData.adresser.map((address: Address) => {
+                      return {
+                          label: `${address.adressetekst} (${address.kommunenavn})`,
+                          lat: address.representasjonspunkt.lat,
+                          lon: address.representasjonspunkt.lon,
+                      };
+                  })
+                : [];
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]: Unnecessary as mapping an empty list returns an empty list.

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        ).then((res) => res.json());
+        const locationList =
+            locationData.adresser.length > 0
+                ? locationData.adresser.map((address: Address) => {
+                      return {
+                          label: `${address.adressetekst} (${address.kommunenavn})`,
+                          lat: address.representasjonspunkt.lat,
+                          lon: address.representasjonspunkt.lon,
+                      };
+                  })
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]:
```suggestion
                ? locationData.adresser.map((address: Address) => ({
                      label: `${address.adressetekst} (${address.kommunenavn})`,
                      lat: address.representasjonspunkt.lat,
                      lon: address.representasjonspunkt.lon,
                  }))
```

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        );
+    };
+
+    const onInputChange = async (e: React.ChangeEvent<HTMLInputElement>): Promise<void> => {
+        const locationData = await fetch(
+            `https://ws.geonorge.no/adresser/v1/sok?sok=${e.target.value}&treffPerSide=10`,
+        ).then((res) => res.json());
+        const locationList =
+            locationData.adresser.length > 0
+                ? locationData.adresser.map((address: Address) => {
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]:
```suggestion
                ? locationData.adresser.map((address) => {
```

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                      };
+                  })
+                : [];
+        setLocations(locationList);
+    };
+
+    const selectionChanged = (address: { label: string; lat: number; lon: number }): void => {
+        if (address === null) return;
+        setState({ lat: address.lat.toString(), lon: address.lon.toString() });
+    };
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]:
```suggestion
    const selectionChanged = (address: { label: string; lat: number; lon: number } | null): void => {
        if (address === null) return;
        setState({ lat: address.lat.toString(), lon: address.lon.toString() });
    };
```

Comment (FredrikAugust) [reply src/components/atoms/LatLonFromLocationSearch.tsx]: And replace the inline type with the extracted type (ref comment above).

File: src/components/atoms/LatLonFromLocationSearch.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    return (
+        <>
+            <Combobox
+                items={locations}
+                placeholder="Adresse"
+                onInput={debounce(onInputChange, 500)}
+                itemToString={(item) => (item ? item.label : '')}
```
Comment (FredrikAugust) [inline src/components/atoms/LatLonFromLocationSearch.tsx]:
```suggestion
                itemToString={(item) => (item?.label ?? '')}
```

**Review by FredrikAugust — CHANGES_REQUESTED** [review]

File: src/components/atoms/LatLonFromLocationSearch.tsx (hunk 5 above)
Comment (Kpaubert) [reply src/components/atoms/LatLonFromLocationSearch.tsx]: Okay!

File: src/components/atoms/LatLonFromLocationSearch.tsx (hunk 6 above)
Comment (Kpaubert) [reply src/components/atoms/LatLonFromLocationSearch.tsx]: ok...ay?

**Review by FredrikAugust — APPROVED** [review]

**Pull request merged**